File: AOC_2022/Day 13/13.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
    public:
    vector<pair<string, string>> packets;
    void init(ifstream& input){
        string line;
        while (getline(input, line))
        {
            string temp = line;
            getline(input, line);
            packets.push_back({temp, line});
            getline(input, line);
        }
    }

    vector<string> helper(string s){
        vector<string> result;
        string temp = "";
        int depth = 0;
        for(int i = 1; i < s.size()-1; i++){
            if(s[i] == '['){
                depth++;
            }
            else if(s[i] == ']'){
                depth--;
            }
            else if(s[i] == ','){
                if(depth == 0){
                    result.push_back(temp);
                    temp = "";
                }
            }
            if(s[i] != ',' || (s[i] == ',' && depth != 0)){
                temp += s[i];
            }
        }
        if(temp != "") result.push_back(temp);
        return result;
    }

    int compare(vector<string> left, vector<string> right){
        for(int i = 0; i < min(left.size(), right.size()); i++){
            // if both left and right are integers
            if(left[i][0] != '[' && right[i][0] != '['){
                if(stoi(left[i]) < stoi(right[i])){
                    return -1;
                }
                else if(stoi(left[i]) > stoi(right[i])){
                    return 1;
                }
            }
            // else if left is a list
            else if(left[i][0] == '[' && right[i][0] != '['){
                int temp = compare(helper(left[i]), helper( "[" + right[i] + "]"));
                if(temp != 0){
                    return temp;
                }
            }
            // else if right is a list
            else if(left[i][0] != '[' && right[i][0] == '['){
                int temp = compare(helper("[" + left[i] + "]"), helper(right[i]));
                if(temp != 0){
                    return temp;
                }
            }
            // else if both left and right are lists
            else{
                int temp = compare(helper(left[i]), helper(right[i]));
                if(temp != 0){
                    return temp;
                }
            }
        }
        return left.size() - right.size();
    }

    void part1(){
        int result = 0;
        for(int i = 0; i < packets.size();i++){
            if(compare(helper(packets[i].first), helper(packets[i].second)) < 0){
                result += (i+1);
            }
        }
        cout<<"Part1: "<<result<<"\n";
    }

    void part2(){
        vector<vector<string>> allPackets;
        for(auto it : packets){
            allPackets.push_back(helper(it.first));
            allPackets.push_back(helper(it.second));
        }
        // adding divider packets 
        allPackets.push_back(helper("[[2]]"));
        allPackets.push_back(helper("[[6]]"));
        // sorting all packets based on the compare function
        sort(allPackets.begin(), allPackets.end(), [&](vector<string> left, vector<string> right){
            return compare(left, right) < 0;
        });
        int result = 1;
        for(int i = 0; i < allPackets.size(); i++){
            if(allPackets[i].empty()) continue;
            if((allPackets[i][0] == "[2]" || allPackets[i][0] == "[6]") && allPackets[i].size() == 1){
                result *= (i+1);
            }
        }
        cout<<"Part2: "<<result<<"\n";
    }

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << "\n";
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}

File: AOC_2022/Day 13/13.py
from functools import cmp_to_key

class Solution:
    def __init__(self):
        self.packets = []

    def init(self, input_file):
        with open(input_file, 'r') as input:
            lines = input.readlines()
            for i in range(0, len(lines), 3):
                packet1 = lines[i].strip()
                packet2 = lines[i + 1].strip()
                self.packets.append((packet1, packet2))

    def helper(self, s):
        result = []
        temp = ""
        depth = 0
        for i in range(1, len(s)-1):
            if s[i] == '[':
                depth += 1
            elif s[i] == ']':
                depth -= 1
            elif s[i] == ',':
                if depth == 0:
                    result.append(temp)
                    temp = ""
            if s[i] != ',' or (s[i] == ',' and depth != 0):
                temp += s[i]
        if temp != "":
            result.append(temp)
        return result

    def compare(self, left, right):
        for l, r in zip(left, right):
            if l[0] != '[' and r[0] != '[':
                if int(l) < int(r):
                    return -1
                elif int(l) > int(r):
                    return 1
            elif l[0] == '[' and r[0] != '[':
                temp = self.compare(self.helper(l), self.helper("[" + r + "]"))
                if temp != 0:
                    return temp
            elif l[0] != '[' and r[0] == '[':
                temp = self.compare(self.helper("[" + l + "]"), self.helper(r))
                if temp != 0:
                    return temp
            else:
                temp = self.compare(self.helper(l), self.helper(r))
                if temp != 0:
                    return temp
        return len(left) - len(right)

    def part1(self):
        result = 0
        for i, packet in enumerate(self.packets):
            if self.compare(self.helper(packet[0]), self.helper(packet[1])) < 0:
                result += (i + 1)
        print("Part1:", result)

    def part2(self):
        all_packets = []
        for packet in self.packets:
            all_packets.extend([self.helper(packet[0]), self.helper(packet[1])])

        # adding divider packets
        all_packets.extend([self.helper("[[2]]"), self.helper("[[6]]")])

        # sorting all packets based on the compare function
        all_packets = sorted(all_packets, key=cmp_to_key(self.compare))

        #print(all_packets)

        result = 1
        for i, packet in enumerate(all_packets):
            if not packet:
                continue
            if (packet[0] == "[2]" or packet[0] == "[6]") and len(packet) == 1:
                result *= (i + 1)
        print("Part2:", result)


if __name__ == "__main__":
    input_file = "input.txt"
    s = Solution()
    s.init(input_file)
    s.part1()
    s.part2()
